# Two-Sample T-Test

In [2]:
pip install pingouin

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 185 kB 6.8 MB/s 
     |████████████████████████████████| 9.9 MB 52.7 MB/s 
  Created wheel for pingouin: filename=pingouin-0.5.2-py3-none-any.whl size=196206 sha256=d471d0c6fcbb8cb674844caa92f4436cd7172c802e44235fca7298b9854e3f2b
  Stored in directory: /root/.cache/pip/wheels/11/5a/63/a6d32fc26fa462c731f65480bfb98ff7bd39b8ebcb4bc6c2fe
  Created wheel for littleutils: filename=littleutils-0.2.2-py3-none-any.whl size=7048 sha256=79abc03d8053dcfb857065f926aa410965573ce1b179d37d04ba4fe1fb91320f
  Stored in directory: /root/.cache/pip/wheels/d6/64/cd/32819b511a488e4993f2fab909a95330289c3f4e0f6ef4676d
Successfully built pingouin littleutils
  Attempting uninstall: statsmodels
    Found existing installation: statsmodels 0.12.2
    Uninstalling statsmodels-0.12.2:
      Successfully uninstalled statsmodels-0.12.2


In [3]:
from statsmodels.stats.weightstats import ttest_ind
import numpy as np
import pandas as pd
import pingouin as pg

In [4]:
data_group1 = np.array([160,150,160,156.12,163.24,160.56,168.56,174.12,167,165])
data_group2 = np.array([164,153,168,166.12,173.24,150.56,158.56,171,157,170])

In [5]:
data_group1

array([160.  , 150.  , 160.  , 156.12, 163.24, 160.56, 168.56, 174.12,
       167.  , 165.  ])

In [7]:
# Conductimg two sample ttest
result = pg.ttest(data_group1,data_group2,correction = True)
result

T        dof alternative  ...   cohen-d   BF10     power
T-test -0.20894  17.574485   two-sided  ...  0.093441  0.404  0.054505

[1 rows x 8 columns]

Group 1 and Group 2 heights are statistically equivalent

In [8]:
data_group1 = np.array([83,87,89,91,98,93,78,63,99,89])
data_group2 = np.array([56,64,54,63,99,33,53,58,29,88])

In [10]:
# Conductimg two sample ttest
result = pg.ttest(data_group1,data_group2,correction = True,confidence= 0.90)
result

T        dof alternative  ...   cohen-d    BF10     power
T-test  3.618761  13.114891   two-sided  ...  1.618359  17.661  0.927838

[1 rows x 8 columns]

# Two way Anova

In [1]:
pip install statsmodels

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [7]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
import pandas as pd
import numpy as np

In [16]:
dataframe = pd.DataFrame({'Fertilizers':np.repeat(['daily','weekly'],15),'Watering': np.repeat(['daily','weekly'],15),
                          'Height':[14,16,15,15,16,13,12,11,
                                    14,15,16,16,17,18,14,13,
                                    14,14,14,15,16,16,17,18,
                                    14,13,14,14,14,15]})

In [17]:
dataframe

Fertilizers Watering  Height
0        daily    daily      14
1        daily    daily      16
2        daily    daily      15
3        daily    daily      15
4        daily    daily      16
5        daily    daily      13
6        daily    daily      12
7        daily    daily      11
8        daily    daily      14
9        daily    daily      15
10       daily    daily      16
11       daily    daily      16
12       daily    daily      17
13       daily    daily      18
14       daily    daily      14
15      weekly   weekly      13
16      weekly   weekly      14
17      weekly   weekly      14
18      weekly   weekly      14
19      weekly   weekly      15
20      weekly   weekly      16
21      weekly   weekly      16
22      weekly   weekly      17
23      weekly   weekly      18
24      weekly   weekly      14
25      weekly   weekly      13
26      weekly   weekly      14
27      weekly   weekly      14
28      weekly   weekly      14
29      weekly   weekly      15

In [22]:
# Performing two way Anova
model = ols('Height ~ C(Fertilizers) + C(Watering)',
            data = dataframe).fit()
result = sm.stats.anova_lm(model,type = 2)

result

df     sum_sq   mean_sq         F    PR(>F)
C(Fertilizers)   1.0   0.033333  0.033333  0.012069  0.913305
C(Watering)      1.0   0.000369  0.000369  0.000133  0.990865
Residual        28.0  77.333333  2.761905       NaN       NaN


*   p-value for Fertilizers is 0.913305

*   p-value for Watering is 0.990865


*   Statistically Fertilizers and Watering have least significance on Height of plant which may infer there can be other factors like Weather, Geography , Temperature which can affect plant height




# Dimensionality Reduction Algorithm

In [23]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [24]:
dataset = pd.read_csv('Wine.csv')

In [25]:
dataset

Alcohol  Malic_Acid   Ash  Ash_Alcanity  Magnesium  Total_Phenols  \
0      14.23        1.71  2.43          15.6        127           2.80   
1      13.20        1.78  2.14          11.2        100           2.65   
2      13.16        2.36  2.67          18.6        101           2.80   
3      14.37        1.95  2.50          16.8        113           3.85   
4      13.24        2.59  2.87          21.0        118           2.80   
..       ...         ...   ...           ...        ...            ...   
173    13.71        5.65  2.45          20.5         95           1.68   
174    13.40        3.91  2.48          23.0        102           1.80   
175    13.27        4.28  2.26          20.0        120           1.59   
176    13.17        2.59  2.37          20.0        120           1.65   
177    14.13        4.10  2.74          24.5         96           2.05   

     Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity   Hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     OD280  Proline  Customer_Segment  
0     3.92     1065                 1  
1     3.40     1050                 1  
2     3.17     1185                 1  
3     3.45     1480                 1  
4     2.93      735                 1  
..     ...      ...               ...  
173   1.74      740                 3  
174   1.56      750                 3  
175   1.56      835                 3  
176   1.62      840                 3  
177   1.60      560                 3  

[178 rows x 14 columns]

In [26]:
X = dataset.iloc[:,:-1].values
X

array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
        1.185e+03],
       ...,
       [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
        8.350e+02],
       [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
        8.400e+02],
       [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
        5.600e+02]])

In [28]:
y = dataset[['Customer_Segment']].values
y

array([[1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
       [2],
    

In [29]:
dataset.isna().sum()

Alcohol                 0
Malic_Acid              0
Ash                     0
Ash_Alcanity            0
Magnesium               0
Total_Phenols           0
Flavanoids              0
Nonflavanoid_Phenols    0
Proanthocyanins         0
Color_Intensity         0
Hue                     0
OD280                   0
Proline                 0
Customer_Segment        0
dtype: int64

In [30]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Alcohol               178 non-null    float64
 1   Malic_Acid            178 non-null    float64
 2   Ash                   178 non-null    float64
 3   Ash_Alcanity          178 non-null    float64
 4   Magnesium             178 non-null    int64  
 5   Total_Phenols         178 non-null    float64
 6   Flavanoids            178 non-null    float64
 7   Nonflavanoid_Phenols  178 non-null    float64
 8   Proanthocyanins       178 non-null    float64
 9   Color_Intensity       178 non-null    float64
 10  Hue                   178 non-null    float64
 11  OD280                 178 non-null    float64
 12  Proline               178 non-null    int64  
 13  Customer_Segment      178 non-null    int64  
dtypes: float64(11), int64(3)
memory usage: 19.6 KB


In [32]:
dataset.shape

(178, 14)

In [33]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 7)

In [34]:
X_train

array([[1.296e+01, 3.450e+00, 2.350e+00, ..., 6.800e-01, 1.750e+00,
        6.750e+02],
       [1.371e+01, 5.650e+00, 2.450e+00, ..., 6.400e-01, 1.740e+00,
        7.400e+02],
       [1.412e+01, 1.480e+00, 2.320e+00, ..., 1.170e+00, 2.820e+00,
        1.280e+03],
       ...,
       [1.237e+01, 1.170e+00, 1.920e+00, ..., 1.120e+00, 3.480e+00,
        5.100e+02],
       [1.305e+01, 2.050e+00, 3.220e+00, ..., 1.130e+00, 3.200e+00,
        8.300e+02],
       [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
        8.350e+02]])

In [35]:
X_test

array([[1.287000e+01, 4.610000e+00, 2.480000e+00, 2.150000e+01,
        8.600000e+01, 1.700000e+00, 6.500000e-01, 4.700000e-01,
        8.600000e-01, 7.650000e+00, 5.400000e-01, 1.860000e+00,
        6.250000e+02],
       [1.328000e+01, 1.640000e+00, 2.840000e+00, 1.550000e+01,
        1.100000e+02, 2.600000e+00, 2.680000e+00, 3.400000e-01,
        1.360000e+00, 4.600000e+00, 1.090000e+00, 2.780000e+00,
        8.800000e+02],
       [1.282000e+01, 3.370000e+00, 2.300000e+00, 1.950000e+01,
        8.800000e+01, 1.480000e+00, 6.600000e-01, 4.000000e-01,
        9.700000e-01, 1.026000e+01, 7.200000e-01, 1.750000e+00,
        6.850000e+02],
       [1.277000e+01, 2.390000e+00, 2.280000e+00, 1.950000e+01,
        8.600000e+01, 1.390000e+00, 5.100000e-01, 4.800000e-01,
        6.400000e-01, 9.899999e+00, 5.700000e-01, 1.630000e+00,
        4.700000e+02],
       [1.243000e+01, 1.530000e+00, 2.290000e+00, 2.150000e+01,
        8.600000e+01, 2.740000e+00, 3.150000e+00, 3.900000e-01,
        1.77

In [36]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [37]:
X_train

array([[-1.09781076e-01,  9.86825853e-01, -5.52071466e-02, ...,
        -1.27381132e+00, -1.23657957e+00, -2.99236776e-01],
       [ 8.25553693e-01,  2.93922862e+00,  2.99517506e-01, ...,
        -1.45457503e+00, -1.25082195e+00, -9.96013989e-02],
       [ 1.33687003e+00, -7.61462077e-01, -1.61624542e-01, ...,
         9.40544142e-01,  2.87355054e-01,  1.55890789e+00],
       ...,
       [-8.45577762e-01, -1.03657338e+00, -1.58052315e+00, ...,
         7.14589503e-01,  1.22735211e+00, -8.06003503e-01],
       [ 2.45909611e-03, -2.55612270e-01,  3.03089733e+00, ...,
         7.59780430e-01,  8.28565481e-01,  1.76816816e-01],
       [ 2.76823962e-01,  1.72341417e+00, -3.74459333e-01, ...,
        -1.68052967e+00, -1.50718478e+00,  1.92173383e-01]])

## Applying PCA

In [38]:
from sklearn.decomposition import PCA
pca = PCA(n_components = 2)
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

In [39]:
X_train

array([[-2.5880849 ,  0.7691162 ],
       [-3.24183176,  2.47365748],
       [ 1.69401449,  0.20435537],
       [-2.02774176,  2.25904967],
       [ 1.10140645, -0.43727299],
       [ 2.46559226, -0.11899176],
       [-2.66360746,  2.43433941],
       [-3.11927321,  0.50511845],
       [-0.6178578 , -2.20386288],
       [ 1.01925385, -0.16852906],
       [ 2.14741735,  0.41190648],
       [ 1.79554303, -2.07489315],
       [-0.74696061, -2.42360501],
       [ 0.73268862, -0.89261225],
       [ 0.72659105, -1.30982256],
       [ 1.1048941 , -2.22364617],
       [ 3.05909362,  1.70495435],
       [ 1.92606944,  1.38925871],
       [-1.72869353, -1.52260043],
       [-1.34931997, -1.75622967],
       [-1.79680052, -1.25011274],
       [ 2.7379988 ,  1.45009048],
       [-2.38662165,  0.45602457],
       [ 2.6781174 ,  0.8809642 ],
       [-2.19389265,  2.82188358],
       [ 1.99532546, -1.59578586],
       [-2.5026475 ,  1.96535854],
       [ 2.52975854,  0.55377032],
       [-0.83061144,

In [40]:
# Model Fit
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 7)
classifier.fit(X_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(random_state=7)

In [41]:
from sklearn.metrics import confusion_matrix,accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test,y_pred)
print(cm)
accuracy_score(y_test,y_pred)

[[ 7  0  0]
 [ 2 15  0]
 [ 0  0 12]]


0.9444444444444444